src/viztracer/modules/vcompressor/vc_dump.c

```diff
@@ -146,9 +147,296 @@ int dump_parsed_trace_events(PyObject* trace_events, FILE* fptr)
         }
     }
 
+    // Iterate through counter events
+    ppos = 0;
+    while (PyDict_Next(counter_events, &ppos, &key, &value)) {
+        uint64_t pid = PyLong_AsLong(PyTuple_GetItem(key, 0));
+        uint64_t tid = PyLong_AsLong(PyTuple_GetItem(key, 1));
+        const char* name = PyUnicode_AsUTF8(PyTuple_GetItem(key, 2));
+        fputc(VC_HEADER_COUNTER_EVENTS, fptr);
+        fwrite(&pid, sizeof(uint64_t), 1, fptr);
+        fwrite(&tid, sizeof(uint64_t), 1, fptr);
+        fwritestr(name, fptr);
+        if (diff_and_write_counter_args(value, fptr) != 0) {
+            goto clean_exit;
+        }
+    }
+clean_exit:
+    if (PyErr_Occurred()) {
+        return 1;
+    }
+
+    return 0;
+}
+
+int diff_and_write_counter_args(PyObject* counter_args, FILE* fptr) {
+    /* there may be several args in a counter, log them all may take more spaces
+    *  so this step is to do diffing between two contiguous timestamp
+    *  and finally we only log those changed args
+    *  Here's a counter_args example that we parsed
+    *  {
+    *      1.1: {"a": 20, "b": 10}
+    *      2.2: {"a": 30, "b": 10}
+    *  }
+    *  In this case, "b" value is not changed, so we only need to log
+    *  {
+    *      1.1: {"a": 20, "b": 10}
+    *      2.2: {"a": 30}
+    *  }
+    */
+    PyObject* cached_args_dict = PyDict_New();
+    PyObject* diffed_args = PyDict_New();
+    PyObject* ts_keys = PyDict_Keys(counter_args);
+    PyObject* ts = NULL;
+    PyObject* cur_diffed_arg = NULL;
+    PyObject* cur_counter_arg = NULL;
+    PyObject* arg_key_list = NULL;
+    PyObject* counter_arg_key = NULL;
+    PyObject* counter_arg_value = NULL;
+    PyObject* cached_arg_value = NULL;
+    PyObject* overflowed_num_string = NULL;
+    Py_ssize_t ppos = 0;
+    uint64_t ts_key_count = 0;
+    uint64_t arg_nums = 0;
+    // sort the args by timestamp so we can diff
+    if (!ts_keys || !PyList_Check(ts_keys)) {
+        PyErr_SetString(PyExc_ValueError, "failed to get timestamp list");
+        goto clean_exit;
+    }
+    ts_key_count = PyList_GET_SIZE(ts_keys);
+    if (PyList_Sort(ts_keys) == -1) {
+        goto clean_exit;
+    }
+    // do diffing between two timestamps and store the result
+    for (uint64_t i = 0; i < ts_key_count; i++) {
+        ts = PyList_GET_ITEM(ts_keys, i);
+        cur_counter_arg = PyDict_GetItem(counter_args, ts);
+        cur_diffed_arg = PyDict_New();
+        ppos = 0;
+        while (PyDict_Next(cur_counter_arg, &ppos, &counter_arg_key, &counter_arg_value)) {
```

Comment:
Please add extra comments here. This is a really long procedure and it would be nice to know which step is what.

For example, this while loop is to iterate all the counter values at a specific timestamp and update some cache?

That would be really helpful in the future when you forget what's this about.

Reply:
Maybe a comment for what is `cur_diffed_arg` as well. There are a lot of variables in this function, and they have similar names with keyword "cur", "diff", "counter" and "arg" in them. It could confuse people. Let's do more comments.